docs/secure-boot.md

```diff
@@ -1,39 +1,172 @@
+* **Stored in:** Encrypted in the EFI partition (`balena-luks.enc`), encryption key in the TPM (TPM key slot is indicated by `balena-luks.ctx` in the EFI partition)
+* **Updated by:** `balenaOS` installer during provisioning
+* **Verified by:** `cryptsetup` initrd script (verified by GRUB as a part of kernel+initrd bundle), encryption key locked to PCRs 0,1,2,3
+* **Verifies:** Nothing, only used for encryption
+
+*Note:* When the installer exits, the TPM key is only protected using PCRs 0,2,3. PCR 1 contains UEFI configuration, which the installer tampers with (by changing the boot order). The system is only fully locked to PCRs 0,1,2,3 on first boot after provisioning.
+
+## boot and EFI partition split
+
+On regular `balenaOS` devices there is a single `resin-boot` or `balena-boot` partition mounted under `/mnt/boot`. This holds both the files necessary to boot the device (e.g. GRUB), as well as files necessary for setting up `balenaOS` (e.g. `config.json`, `system-connections`). With secure boot enabled the single boot partition is split in two:
+* The `balena-efi` partition is the only one that stays unencrypted. It contains GRUB, 2nd stage linux kernel, the encrypted passphrase and the metadata necessary to talk to the TPM.
+* The `balena-boot` partition is encrypted and contains everything else, as these files may contain secrets such as passwords or API keys which the encryption should protect.
+
+The partitions are mounted under `/mnt/boot` and `/mnt/efi` respectively and `/mnt/boot/EFI` is symlinked to `../efi/EFI` to make sure all the paths are still valid.
+
+## Using linux kernel as a 2nd stage bootloader
+
+In order to avoid the need to make GRUB use the TPM to unlock the partitions when looking for the kernel, we keep an unencrypted copy of the linux kernel in the EFI partition whose sole purpose is to unlock the root partition using the TPM, find the actual kernel and kexec into it. This is a simple version of a completely different feature but it roughly works as follows:
+1. During provisioning the installer copies an instance of its own/currently running kernel into the EFI partition.
+2. When the newly provisioned OS boots, UEFI loads GRUB.
+3. GRUB finds the unencrypted 2nd stage kernel in the EFI partition and executes it adding `balena_stage2` parameter to the kernel command line.
+4. The attached initramfs uses the TPM to unlock the root partition.
+5. Next the initramfs checks whetner the `balena_stage2` parameter is present and if it is, loads the actual kernel from the root partition and kexec's into it, removing the `balena_stage2` parameter from the kernel command line and replacing the `root` parameter by the actual UUID (which is only available after unlocking LUKS).
+6. The actual kernel from the root partition is executed, boot continues as usual.
+
+Though not currently implemented, the 2nd stage kernel in the EFI partition is intended to be replaced by currently running kernel on successful HUP.
```

Comment:
There is a task for this, I would not mention it on the documentation unless it was technical debt or a known issue.